storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1760 @@ buf_block_t *fsp_get_latched_xdes_page(
+
+    uint32_t parent_offset= uint32_t(parent_it->second >> 32);
+
+    if (parent_offset > srv_page_size - FIL_PAGE_DATA_END)
+    {
+      err= DB_CORRUPTION;
+      goto err_exit;
+    }
+
+    pageAllocator allocator(header_block, iblock, inode, extent_size,
+                            &mtr);
+
+    associatedPages related_pages(block, &mtr);
+    err= allocator.prepare(new_extent, old_state == XDES_FSEG);
+
+    DBUG_EXECUTE_IF("allocation_prepare_fail", err= DB_CORRUPTION;);
+    if (err) goto err_exit;
+
+    err= related_pages.prepare(parent_page_no);
+    DBUG_EXECUTE_IF("relation_page_prepare_fail", err= DB_CORRUPTION;);
+
+    if (err) goto err_exit;
+
+    allocator.complete();
+
+    /* After allocating the new page, try to prepare the steps
+    of page removal function. Because there is a possiblity that
+    last block in FSEG_NOT_FULL/FSP_FREE_FRAG/FSP_FREE last block
+    could've changed while allocating the new block. */
+    pageRemover remover(header_block, iblock, inode, extent_size,
+                        cur_page_no, &mtr);
+    err= remover.prepare();
+ 
+    DBUG_EXECUTE_IF("remover_prepare_fail", err= DB_CORRUPTION;);
+    if (err)
+    {
+      allocator.restore_old_pages();
+      mtr.discard_modifications();
+      goto err_exit;
+    }
+
+    /* Copy the data from old block to new block */
+    buf_block_t *new_block= allocator.get_new_block();
+    uint32_t new_page_no= new_block->page.id().page_no();
+    mtr.memcpy(*new_block, new_block->page.frame,
+               block->page.frame, srv_page_size - FIL_PAGE_DATA_END);
+    mtr.write<4,mtr_t::FORCED>(*new_block,
+                               FIL_PAGE_OFFSET + new_block->page.frame,
+                               new_page_no);
+
+    /* Assign the new block page number in left, right
+    and parent block */
+    related_pages.complete(new_page_no, parent_offset);
+
+    /* Complete the page free operation */
+    remover.complete(); 
+    /* Add the new page in inode fragment array */
+    allocator.assign_frag_slot();
+
+    if (level)
+    {
+      err= store_child_pages(new_block);
+      if (err) goto err_exit;
+    }
+    goto fetch_next_page;
+  }
+
+  ut_ad(!fsp_tablespace_validate(fil_system.sys_space));
+  ut_ad(!fseg_validate(fil_system.sys_space, m_index));
+  return DB_SUCCESS;
+}
+
+dberr_t indexDefragmenter::defragment(spaceDefragmenter *space_defrag) noexcept
+{
+  mtr_t mtr;
+  dberr_t err= DB_SUCCESS;
+  mtr.start();
+  mtr.x_lock_space(fil_system.sys_space);
+  buf_block_t *root= buf_page_get_gen(page_id_t{0, m_index->page},
+                                      0, RW_SX_LATCH, nullptr,
+                                      BUF_GET_POSSIBLY_FREED, &mtr,
+                                      &err);
+  if (err)
+  {
+    mtr.commit();
+    return err;
+  }
+  uint16_t level= btr_page_get_level(root->page.frame);
+  mtr.commit();
+  while (1)
+  {
+    err= defragment_level(level, space_defrag);
```

Comment:
It seems to me that we could use the same `mtr` object for `defragment_level()` in order to avoid allocating multiple `mtr_t` objects.